sql: TestMigrateSchemaChanges/drop-table-blocked-at-after-reversing-mutations-canceled is flaky

On 82294d100ad1cb26b488f53e7241242d474d58d7, running
```
make roachprod-stress CLUSTER=lucy-stress STRESSFLAGS='-p 16' PKG=./pkg/sql TESTS=TestMigrateSchemaChanges
```
produces
```
    --- FAIL: TestMigrateSchemaChanges/drop-table-blocked-at-after-reversing-mutations-canceled (48.59s)
        schema_change_migrations_test.go:202: condition failed to evaluate within 45s: query 'SELECT * FROM [SHOW JOBS] WHERE (job_type = 'SCHEMA CHANGE' OR job_type = 'SCHEMA CHANGE GC') AND NOT (status = 'succeeded' OR status = 'canceled')': expected:

            got:
            546898560510820353, SCHEMA CHANGE, DROP TABLE t.public.test, , root, failed, NULL, 2020-04-15 17:09:55.405294 +0000 +0000, 2020-04-15 17:09:55.519922 +0000 +0000, 2020-04-15 17:09:56.827307 +0000 +0000, 2020-04-15 17:09:56.353098 +0000 +0000, 0, job 546898560510820353: cannot be reverted, manual cleanup may be required: job canceled by user, 2


            goroutine 84664 [running]:
            runtime/debug.Stack(0xc005ef9ce0, 0x46c6160, 0xc008db9460)
            	/usr/local/go/src/runtime/debug/stack.go:24 +0x9d
            github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon(0x47b8960, 0xc006f49b00, 0xc005ef9ce0)
            	/go/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:37 +0x6b
            github.com/cockroachdb/cockroach/pkg/testutils/sqlutils.(*SQLRunner).CheckQueryResultsRetry(0xc00415ba30, 0x47b8960, 0xc006f49b00, 0x3fbeeff, 0x93, 0x6b07f10, 0x0, 0x0)
            	/go/src/github.com/cockroachdb/cockroach/pkg/testutils/sqlutils/sql_runner.go:199 +0xde
            github.com/cockroachdb/cockroach/pkg/sql_test.testSchemaChangeMigrations(0xc006f49b00, 0x2, 0x1, 0x7, 0x3efbc7e, 0x18)
            	/go/src/github.com/cockroachdb/cockroach/pkg/sql/schema_change_migrations_test.go:202 +0x71b
            github.com/cockroachdb/cockroach/pkg/sql_test.TestMigrateSchemaChanges.func1(0xc006f49b00)
            	/go/src/github.com/cockroachdb/cockroach/pkg/sql/schema_change_migrations_test.go:846 +0x92
            testing.tRunner(0xc006f49b00, 0xc00de56e40)
            	/usr/local/go/src/testing/testing.go:909 +0xc9
            created by testing.(*T).Run
            	/usr/local/go/src/testing/testing.go:960 +0x350
```
This took about half an hour and ~700 runs. It'll be faster to reproduce without running the other subtests.

This seems to be fallout from #47446, which is uncovering a pre-existing problem with the tests (and probably doesn't reflect a problem with #47446 itself). The problem is that we were expecting a canceled job, but in some cases rolling back the job was failing. This error would have been swallowed before #47446, but now it's getting propagated properly and causing the job to be failed instead of canceled.

Full logs: [out.1586968821.gz](https://github.com/cockroachdb/cockroach/files/4482737/out.1586968821.gz)

--- a/pkg/sql/schema_changer.go
+++ b/pkg/sql/schema_changer.go
@@ -636,7 +636,7 @@ func (sc *SchemaChanger) initJobRunningStatus(ctx context.Context) error {
 func (sc *SchemaChanger) rollbackSchemaChange(ctx context.Context, err error) error {
 	log.Warningf(ctx, "reversing schema change %d due to irrecoverable error: %s", *sc.job.ID(), err)
 	if errReverse := sc.maybeReverseMutations(ctx, err); errReverse != nil {
-		return err
+		return errReverse
 	}
 
 	if fn := sc.testingKnobs.RunAfterMutationReversal; fn != nil {
@@ -1036,6 +1036,12 @@ func (sc *SchemaChanger) refreshStats() {
 // applying a schema change. If a column being added is reversed and dropped,
 // all new indexes referencing the column will also be dropped.
 func (sc *SchemaChanger) maybeReverseMutations(ctx context.Context, causingError error) error {
+	if fn := sc.testingKnobs.RunBeforeMutationReversal; fn != nil {
+		if err := fn(*sc.job.ID()); err != nil {
+			return err
+		}
+	}
+
 	// Get the other tables whose foreign key backreferences need to be removed.
 	var fksByBackrefTable map[sqlbase.ID][]*sqlbase.ConstraintToUpdate
 	alreadyReversed := false
@@ -1431,6 +1437,9 @@ type SchemaChangerTestingKnobs struct {
 	// RunBeforeConstraintValidation is called just before starting the checks validation,
 	// after setting the job status to validating.
 	RunBeforeConstraintValidation func() error
+
+	// RunBeforeMutationReversal runs at the beginning of maybeReverseMutations.
+	RunBeforeMutationReversal func(jobID int64) error
 
 	// RunAfterMutationReversal runs in OnFailOrCancel after the mutations have
 	// been reversed.

--- a/pkg/sql/schema_changer_test.go
+++ b/pkg/sql/schema_changer_test.go
@@ -5811,54 +5811,84 @@ func TestRetriableErrorDuringRollback(t *testing.T) {
 	defer setTestJobsAdoptInterval()()
 	ctx := context.Background()
 
-	onFailOrCancelStarted := false
-	injectedError := false
-	params, _ := tests.CreateTestServerParams()
-	params.Knobs = base.TestingKnobs{
-		SQLSchemaChanger: &sql.SchemaChangerTestingKnobs{
-			RunBeforeOnFailOrCancel: func(_ int64) error {
-				onFailOrCancelStarted = true
-				return nil
-			},
-			RunBeforeBackfill: func() error {
-				// The first time through the backfiller in OnFailOrCancel, return a
-				// retriable error.
-				if !onFailOrCancelStarted || injectedError {
-					return nil
-				}
-				injectedError = true
-				// Return an artificial context canceled error.
-				return context.Canceled
-			},
-		},
-	}
-	s, sqlDB, kvDB := serverutils.StartServer(t, params)
-	defer s.Stopper().Stop(ctx)
+	runTest := func(params base.TestServerArgs) {
+		s, sqlDB, kvDB := serverutils.StartServer(t, params)
+		defer s.Stopper().Stop(ctx)
 
-	// Disable strict GC TTL enforcement because we're going to shove a zero-value
-	// TTL into the system with addImmediateGCZoneConfig.
-	defer disableGCTTLStrictEnforcement(t, sqlDB)()
+		// Disable strict GC TTL enforcement because we're going to shove a zero-value
+		// TTL into the system with addImmediateGCZoneConfig.
+		defer disableGCTTLStrictEnforcement(t, sqlDB)()
 
-	_, err := sqlDB.Exec(`
+		_, err := sqlDB.Exec(`
 CREATE DATABASE t;
 CREATE TABLE t.test (k INT PRIMARY KEY, v INT8);
 INSERT INTO t.test VALUES (1, 2), (2, 2);
 `)
-	require.NoError(t, err)
-	tableDesc := sqlbase.GetTableDescriptor(kvDB, "t", "test")
-	// Add a zone config for the table.
-	_, err = addImmediateGCZoneConfig(sqlDB, tableDesc.ID)
-	require.NoError(t, err)
+		require.NoError(t, err)
+		tableDesc := sqlbase.GetTableDescriptor(kvDB, "t", "test")
+		// Add a zone config for the table.
+		_, err = addImmediateGCZoneConfig(sqlDB, tableDesc.ID)
+		require.NoError(t, err)
 
-	// Try to create a unique index which won't be valid and will need a rollback.
-	_, err = sqlDB.Exec(`
+		// Try to create a unique index which won't be valid and will need a rollback.
+		_, err = sqlDB.Exec(`
 CREATE UNIQUE INDEX i ON t.test(v);
 `)
-	require.Error(t, err)
-	require.Regexp(t, "violates unique constraint", err.Error())
-	// Verify that the index was cleaned up.
-	testutils.SucceedsSoon(t, func() error {
-		return checkTableKeyCountExact(ctx, kvDB, 2)
+		require.Regexp(t, "violates unique constraint", err.Error())
+		// Verify that the index was cleaned up.
+		testutils.SucceedsSoon(t, func() error {
+			return checkTableKeyCountExact(ctx, kvDB, 2)
+		})
+	}
+
+	t.Run("error-before-backfill", func(t *testing.T) {
+		onFailOrCancelStarted := false
+		injectedError := false
+		params, _ := tests.CreateTestServerParams()
+		params.Knobs = base.TestingKnobs{
+			SQLSchemaChanger: &sql.SchemaChangerTestingKnobs{
+				RunBeforeOnFailOrCancel: func(_ int64) error {
+					onFailOrCancelStarted = true
+					return nil
+				},
+				RunBeforeBackfill: func() error {
+					// The first time through the backfiller in OnFailOrCancel, return a
+					// retriable error.
+					if !onFailOrCancelStarted || injectedError {
+						return nil
+					}
+					injectedError = true
+					// Return an artificial context canceled error.
+					return context.Canceled
+				},
+			},
+		}
+		runTest(params)
+	})
+
+	t.Run("error-before-reversing-mutations", func(t *testing.T) {
+		onFailOrCancelStarted := false
+		injectedError := false
+		params, _ := tests.CreateTestServerParams()
+		params.Knobs = base.TestingKnobs{
+			SQLSchemaChanger: &sql.SchemaChangerTestingKnobs{
+				RunBeforeOnFailOrCancel: func(_ int64) error {
+					onFailOrCancelStarted = true
+					return nil
+				},
+				RunBeforeMutationReversal: func(_ int64) error {
+					// The first time through reversing mutations, return a retriable
+					// error.
+					if !onFailOrCancelStarted || injectedError {
+						return nil
+					}
+					injectedError = true
+					// Return an artificial context canceled error.
+					return context.Canceled
+				},
+			},
+		}
+		runTest(params)
 	})
 }
 
@@ -5870,46 +5900,76 @@ func TestPermanentErrorDuringRollback(t *testing.T) {
 	defer setTestJobsAdoptInterval()()
 	ctx := context.Background()
 
-	onFailOrCancelStarted := false
-	injectedError := false
-	params, _ := tests.CreateTestServerParams()
-	params.Knobs = base.TestingKnobs{
-		SQLSchemaChanger: &sql.SchemaChangerTestingKnobs{
-			RunBeforeOnFailOrCancel: func(_ int64) error {
-				onFailOrCancelStarted = true
-				return nil
-			},
-			RunBeforeBackfill: func() error {
-				// The first time through the backfiller in OnFailOrCancel, return a
-				// permanent error.
-				if !onFailOrCancelStarted || injectedError {
-					return nil
-				}
-				injectedError = true
-				// Any error not on the whitelist of retriable errors is considered permanent.
-				return errors.New("permanent error")
-			},
-		},
-	}
-	s, sqlDB, _ := serverutils.StartServer(t, params)
-	defer s.Stopper().Stop(ctx)
+	runTest := func(params base.TestServerArgs) {
+		s, sqlDB, _ := serverutils.StartServer(t, params)
+		defer s.Stopper().Stop(ctx)
 
-	_, err := sqlDB.Exec(`
+		_, err := sqlDB.Exec(`
 CREATE DATABASE t;
 CREATE TABLE t.test (k INT PRIMARY KEY, v INT8);
 INSERT INTO t.test VALUES (1, 2), (2, 2);
 `)
-	require.NoError(t, err)
+		require.NoError(t, err)
 
-	// Try to create a unique index which won't be valid and will need a rollback.
-	_, err = sqlDB.Exec(`
+		// Try to create a unique index which won't be valid and will need a rollback.
+		_, err = sqlDB.Exec(`
 CREATE UNIQUE INDEX i ON t.test(v);
 `)
-	require.Error(t, err)
-	require.Regexp(t, "violates unique constraint", err.Error())
+		require.Regexp(t, "violates unique constraint", err.Error())
 
-	var jobErr string
-	row := sqlDB.QueryRow("SELECT error FROM [SHOW JOBS] WHERE job_type = 'SCHEMA CHANGE'")
-	require.NoError(t, row.Scan(&jobErr))
-	require.Regexp(t, "cannot be reverted, manual cleanup may be required", jobErr)
+		var jobErr string
+		row := sqlDB.QueryRow("SELECT error FROM [SHOW JOBS] WHERE job_type = 'SCHEMA CHANGE'")
+		require.NoError(t, row.Scan(&jobErr))
+		require.Regexp(t, "cannot be reverted, manual cleanup may be required: permanent error", jobErr)
+	}
+
+	t.Run("error-before-backfill", func(t *testing.T) {
+		onFailOrCancelStarted := false
+		injectedError := false
+		params, _ := tests.CreateTestServerParams()
+		params.Knobs = base.TestingKnobs{
+			SQLSchemaChanger: &sql.SchemaChangerTestingKnobs{
+				RunBeforeOnFailOrCancel: func(_ int64) error {
+					onFailOrCancelStarted = true
+					return nil
+				},
+				RunBeforeBackfill: func() error {
+					// The first time through the backfiller in OnFailOrCancel, return a
+					// permanent error.
+					if !onFailOrCancelStarted || injectedError {
+						return nil
+					}
+					injectedError = true
+					// Any error not on the whitelist of retriable errors is considered permanent.
+					return errors.New("permanent error")
+				},
+			},
+		}
+		runTest(params)
+	})
+
+	t.Run("error-before-reversing-mutations", func(t *testing.T) {
+		onFailOrCancelStarted := false
+		injectedError := false
+		params, _ := tests.CreateTestServerParams()
+		params.Knobs = base.TestingKnobs{
+			SQLSchemaChanger: &sql.SchemaChangerTestingKnobs{
+				RunBeforeOnFailOrCancel: func(_ int64) error {
+					onFailOrCancelStarted = true
+					return nil
+				},
+				RunBeforeBackfill: func() error {
+					// The first time through reversing mutations, return a permanent
+					// error.
+					if !onFailOrCancelStarted || injectedError {
+						return nil
+					}
+					injectedError = true
+					// Any error not on the whitelist of retriable errors is considered permanent.
+					return errors.New("permanent error")
+				},
+			},
+		}
+		runTest(params)
+	})
 }

